service/tablet_allocator.cc

```diff
@@ -12,12 +12,419 @@
+            if (!tmap.transitions().empty()) {
+                // FIXME: The algorithm doesn't support balancing with active transitions yet. They must finish first.
+                lblogger.warn("Pending transitions active.");
+                co_return migration_plan();
+            }
+
+            for (const tablet_info& ti : tmap.tablets()) {
+                co_await coroutine::maybe_yield();
+                for (auto&& replica : ti.replicas) {
+                    if (!nodes.contains(replica.host)) {
+                        continue;
+                    }
+                    auto& node_load_info = nodes[replica.host];
+                    auto&& shard_load_info = node_load_info.shards[replica.shard];
+                    if (shard_load_info.tablet_count == 0) {
+                        node_load_info.candidates.push_back(replica.shard);
+                    }
+                    shard_load_info.tablet_count += 1;
+                    shard_load_info.candidates.emplace(global_tablet_id{table, *tid});
+                }
+                tid = tmap.next_tablet(*tid);
+            }
+        }
+
+        // Prepare candidate nodes and shards for heap-based balancing.
+
+        // heap which tracks most-loaded nodes in terms of avg_load.
+        std::vector<host_id> candidate_nodes;
+        candidate_nodes.reserve(nodes.size());
+        auto nodes_cmp = [&] (const host_id& a, const host_id& b) {
+            return nodes[a].avg_load < nodes[b].avg_load;
+        };
+
+        for (auto&& [host, node_load] : nodes) {
+            if (lblogger.is_enabled(seastar::log_level::debug)) {
+                shard_id shard = 0;
+                for (auto&& shard_load : node_load.shards) {
+                    lblogger.debug("shard {}: all tablets: {}, candidates: {}", tablet_replica{host, shard},
+                                   shard_load.tablet_count, shard_load.candidates.size());
+                    shard++;
+                }
+            }
+
+            candidate_nodes.push_back(host);
+            std::make_heap(node_load.candidates.begin(), node_load.candidates.end(), node_load.candidates_cmp());
+        }
+
+        std::make_heap(candidate_nodes.begin(), candidate_nodes.end(), nodes_cmp);
+
+        locator::load_sketch target_load(_tm);
+        co_await target_load.populate(target);
+        migration_plan plan;
+        const tablet_metadata& tmeta = _tm->tablets();
+        load_type max_off_candidate_load = 0; // max load among nodes which ran out of candidates.
+        while (plan.size() < batch_size && !candidate_nodes.empty()) {
+            co_await coroutine::maybe_yield();
+
+            std::pop_heap(candidate_nodes.begin(), candidate_nodes.end(), nodes_cmp);
+            auto src_host = candidate_nodes.back();
+            auto& src_node_info = nodes[src_host];
+
+            if (std::max(max_off_candidate_load, src_node_info.avg_load) == nodes[target].avg_load) {
```

Comment:
Also please explain in a comment why this equality implies balance. I don't get it.

Reply:
It is possible to have equality, if nodes have the same shard count.

I agree that the condition for balance can be improved, but it doesn't matter that much right now. In the worst case we will stop due to load inversion check. We could avoid building data structures if we had a smarter check, but this doesn't matter now when the load balancer is invoked rarely.

Reply:
Which part is unclear? The goal is to equalize avg_load. max_off_candidate_load is higher than src_node_info.avg_load, so `std::max(max_off_candidate_load, src_node_info.avg_load)` is the max avg_load among non-target nodes. If that is equal to target's avg_load, it means all nodes have the same avg_load.

Reply:
```
            // Check if all nodes reached the same avg_load. There are three sets of nodes: target, candidates (nodes_by_load) 
            // and off-candidates (removed from nodes_by_load). At any time, the avg_load for target is not greater than
            // that of any candidate, and avg_load of any candidate is not greater than that of any in the off-candidates set.
            // This is ensured by the fact that we remove candidates in the order of avg_load from the heap, and
            // because we prevent load inversion between candidate and target in the next check.
            // So the max avg_load of candidates is that of the current src_node_info, and max avg_load of off-candidates
            // is tracked in max_off_candidate_load. If max_off_candidate_load is equal to target's avg_load,
            // it means that all nodes have equal avg_load. We take the maximum with the current candidate in src_node_info
            // to handle the case of off-candidates being empty. In that case, max_off_candidate_load is 0.
```

Reply:
> Is it even possible to achieve equality here in general? Especially since we're talking about doubles... We should at least use `diff < epsilon`.

It's not really important. We'll move to balancing on size (and later on CPU usage too, maybe).

I consider the algorithm here as a napkin sketch, what's more important is the code that drives the algorithm. Even that is just a sketch here as it doesn't work with edges.